### ISSUE TITLE
yang-library option is not working as expected

### ISSUE THREAD

**Issue by ahassany** [CESNET/libyang#2190]
I've tried to generate yang-library using libyang (built from master) and then validate yang data with it but I get strange errors. 

I tried with a simple example (shown below) of ietf-interfaces model.  In search path I put the following yang models: `iana-if-type@2014-05-08.yang`, `ietf-inet-types@2013-07-15.yang`, `ietf-interfaces@2018-02-20.yang`, `ietf-ip@2018-02-22.yang`, and `ietf-yang-types@2013-07-15.yang`.

With normal operation I can validate the example yang data with no problem:
`yanglint -p . iana-if-type@2014-05-08.yang ietf-interfaces@2018-02-20.yang example.json`

Then I proceeded to generate a matching yang-library using
`yanglint -y -p . iana-if-type@2014-05-08.yang ietf-interfaces@2018-02-20.yang -f json -l > lib.json`

That produced a library file (shown below).
However, when I use the generate library file I get a strange error:

```bach
yanglint -y -Y lib.json example.json
libyang err : Mandatory node "content-id" instance does not exist. (Data location "/ietf-yang-library:yang-library".)
libyang err : Mandatory node "module-set-id" instance does not exist. (Data location "/ietf-yang-library:modules-state".)
YANGLINT[E]: Failed to parse input data file "example.json".
```

I checked the generated yang library file and it does contain the `content-id` and `module-set-id` nodes. Could that be a bug or am I missing something here?

Example used:
```json
{
    "ietf-interfaces:interfaces": {
      "interface": [
        {
          "name": "eth0",
          "description": "Wire Connection",
          "type": "iana-if-type:ethernetCsmacd",
          "enabled": true,
          "admin-status": "up",
          "oper-status": "up",
          "if-index": 1,
          "statistics": {
            "discontinuity-time": "2023-02-04T17:30:09.44+01:00"
          }
        }
      ]
    }
  }
```

Generate yang-library "lib.json"
```json
{
  "ietf-yang-library:yang-library": {
    "module-set": [
      {
        "name": "complete",
        "module": [
          {
            "name": "yang",
            "revision": "2022-06-16",
            "namespace": "urn:ietf:params:xml:ns:yang:1"
          },
          {
            "name": "ietf-yang-schema-mount",
            "revision": "2019-01-14",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-schema-mount"
          },
          {
            "name": "ietf-datastores",
            "revision": "2018-02-14",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-datastores"
          },
          {
            "name": "ietf-yang-library",
            "revision": "2019-01-04",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-library"
          },
          {
            "name": "iana-if-type",
            "revision": "2014-05-08",
            "namespace": "urn:ietf:params:xml:ns:yang:iana-if-type",
            "location": [
              "file:///home/user1/repos/libyang/build/example/iana-if-type@2014-05-08.yang"
            ]
          },
          {
            "name": "ietf-interfaces",
            "revision": "2018-02-20",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-interfaces",
            "location": [
              "file:///home/user1/repos/libyang/build/example/ietf-interfaces@2018-02-20.yang"
            ],
            "feature": [
              "arbitrary-names",
              "pre-provisioning",
              "if-mib"
            ]
          }
        ],
        "import-only-module": [
          {
            "name": "ietf-yang-metadata",
            "revision": "2016-08-05",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-metadata"
          },
          {
            "name": "ietf-inet-types",
            "revision": "2013-07-15",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-inet-types"
          },
          {
            "name": "ietf-yang-types",
            "revision": "2013-07-15",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-types"
          },
          {
            "name": "ietf-yang-structure-ext",
            "revision": "2020-06-17",
            "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-structure-ext"
          }
        ]
      }
    ],
    "schema": [
      {
        "name": "complete",
        "module-set": [
          "complete"
        ]
      }
    ],
    "content-id": "17"
  },
  "ietf-yang-library:modules-state": {
    "module-set-id": "17",
    "module": [
      {
        "name": "ietf-yang-metadata",
        "revision": "2016-08-05",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-metadata",
        "conformance-type": "import"
      },
      {
        "name": "yang",
        "revision": "2022-06-16",
        "namespace": "urn:ietf:params:xml:ns:yang:1",
        "conformance-type": "implement"
      },
      {
        "name": "ietf-inet-types",
        "revision": "2013-07-15",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-inet-types",
        "conformance-type": "import"
      },
      {
        "name": "ietf-yang-types",
        "revision": "2013-07-15",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-types",
        "conformance-type": "import"
      },
      {
        "name": "ietf-yang-schema-mount",
        "revision": "2019-01-14",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-schema-mount",
        "conformance-type": "implement"
      },
      {
        "name": "ietf-yang-structure-ext",
        "revision": "2020-06-17",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-structure-ext",
        "conformance-type": "import"
      },
      {
        "name": "ietf-datastores",
        "revision": "2018-02-14",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-datastores",
        "conformance-type": "implement"
      },
      {
        "name": "ietf-yang-library",
        "revision": "2019-01-04",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-library",
        "conformance-type": "implement"
      },
      {
        "name": "iana-if-type",
        "revision": "2014-05-08",
        "schema": "file:///home/user1/repos/libyang/build/example/iana-if-type@2014-05-08.yang",
        "namespace": "urn:ietf:params:xml:ns:yang:iana-if-type",
        "conformance-type": "implement"
      },
      {
        "name": "ietf-interfaces",
        "revision": "2018-02-20",
        "schema": "file:///home/user1/repos/libyang/build/example/ietf-interfaces@2018-02-20.yang",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-interfaces",
        "feature": [
          "arbitrary-names",
          "pre-provisioning",
          "if-mib"
        ],
        "conformance-type": "implement"
      }
    ]
  }
}
```

**Comment by michalvasko** [comment]
The `-Y` parameter only causes the provided file to be used for creating the YANG schema context, it is not included in the validation of the data file. And since the default data type is `data`, which is both state and configuration data, the error is correct.

TLDR; Executing `yanglint -y -Y lib.json -t config example.json` should work as expected.

**Comment by ahassany** [comment]
hmm, I just tried it again. But still doesn't work.
```bassh
❯ yanglint -y -Y lib.json -t config example.json
libyang err : Unexpected data state node "admin-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/admin-status", line number 9.)
libyang err : Unexpected data state node "oper-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status", line number 10.)
libyang err : Unexpected data state node "if-index" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/if-index", line number 11.)
libyang err : Unexpected data state node "statistics" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/statistics", line number 12.)
YANGLINT[E]: Failed to parse input data file "example.json".
```

**Comment by michalvasko** [comment]
I missed that you have actually included state nodes in the JSON file. Then there are 2 solutions, either manually remove `ietf-yang-library` from the `lib.json` file to prevent it from being loaded or use `-e` parameter to force validation of only the modules with some parsed data (only `ietf-interfaces`).

**Comment by ahassany** [comment]
Thank you Michal for the hints, there were very helpful. I've tried different combinations and that what worked and didn't work for me:

With `-e` option, this the output I'm getting:
```bash
❯ yanglint -y -Y lib.json -t config -e example.json
libyang err : Unexpected data state node "admin-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/admin-status", line number 9.)
libyang err : Unexpected data state node "oper-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status", line number 10.)
libyang err : Unexpected data state node "if-index" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/if-index", line number 11.)
libyang err : Unexpected data state node "statistics" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/statistics", line number 12.)
YANGLINT[E]: Failed to parse input data file "example.json".
```

And for the second option for removing the `ietf-yang-library`
```bash
❯ yanglint -Y lib-without-yanglib.json -t config  example.json
libyang err : Unexpected data state node "admin-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/admin-status", line number 9.)
libyang err : Unexpected data state node "oper-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status", line number 10.)
libyang err : Unexpected data state node "if-index" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/if-index", line number 11.)
libyang err : Unexpected data state node "statistics" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/statistics", line number 12.)
YANGLINT[E]: Failed to parse input data file "example.json".

# With -e option
❯ yanglint -Y lib-without-yanglib.json -t config -e example.json
libyang err : Unexpected data state node "admin-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/admin-status", line number 9.)
libyang err : Unexpected data state node "oper-status" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/oper-status", line number 10.)
libyang err : Unexpected data state node "if-index" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/if-index", line number 11.)
libyang err : Unexpected data state node "statistics" found. (Data location "/ietf-interfaces:interfaces/interface[name='eth0']/statistics", line number 12.)
YANGLINT[E]: Failed to parse input data file "example.json".
```

But what really worked was:
```bash
❯ yanglint -Y lib.json -e example.json
# And without the ietf-yang-library
❯ yanglint -Y lib-without-yanglib.json -e example.json
```


Maybe, it's good to document this behavior somewhere, so other people using the yang-library to validate data won't be confused with it. I know what works now, though not why yet 🙈.

**Comment by michalvasko** [comment]
I meant to replace my previous advice with the new one, so you were meant to drop `-t config`, sorry for not mentioning it explicitly.

**Comment by ahassany** [comment]
No worries,  your pointers where helpful indeed.
